content/3_Silver/Flood_Fill.mdx

```diff
@@ -723,6 +723,68 @@ def floodfill(r: int, c: int):
 		frontier.append((r, c + 1))
 
 
+room_num = 0
+for row in range(row_num):
+	for col in range(col_num):
+		if not visited[row][col] and grid[row][col] == '.':
+			floodfill(row, col)
+			room_num += 1
+print(room_num)
+```
+
+</PySection>
+
+</LanguageSection>
+
+<!-- Below is a python specific optimization. -->
+<LanguageSection>
+
+<CPPSection>
+<!-- Add a section for a different language so the text below only shows for python. -->
+</CPPSection>
+
+<PySection>
+
+## Python Optimization: Padding
+
+We can speed up floodfill by a constant time factor by taking advantage of negative indexing in python. Imagine a 1-cell padding on all four sides of the array, for which the cells of which are marked as already having been visited. We can then get rid of the bound checking where we check if the next cell we visit is inclusive between 0 and the dimensions of the array. These cells in the padding may be added to the queue, but when they are visited in the floodfill loop, they are skipped for being marked visited and no out-of-bounds accessing happens. Furthermore, no additional out-of-bounds cells beyond the padding are added to the queue.
+
+Typically, adding padding would need $O(n^2)$ time as we would need to shift the entirety each row to add padding to the front of each row, and also shift each row down to add padding acting as the first row. However, it can be done in linear tine in python as we can have the padding at the right and bottom of the matrix act as the padding for the left and top as the coordinate for the left and top, -1, would be interpreted as the end of the array.
```

Comment:
why would it be n^2? couldn't you add some padding _before_ reading in input?

Reply:
That's a super good point lol

Reply:
Yes you could. I'll rewrite it. I guess python seems like the only language where getting a TLE because of floodfill would be a problem even if floodfill is the intended solution.